### PR TITLE
Expose D frontend via DUB

### PR DESCRIPTION
**Pull request by wilzbach**
Allows to use the entire frontend as a DUB package.

There's still an ugly lot of initialization needed and two hard-coded bits (platform, importPaths), but one can run through all semantic phases which is more than needed if we want to take the DMD as a library project serious ;-)

I will comment on the individual lines for which I would like to find a better solution.

Also note that I didn't add this to the test scripts yet as I am waiting for https://github.com/dlang/dmd/pull/7415 (allows to test the DUB package on Travis).

### CODE REVIEW

**Comment by wilzbach** [comment]
> wilzbach requested a review from WalterBright as a code owner just now

Btw @WalterBright this happens automatically because of

https://github.com/dlang/dmd/blob/f27ee714def99771232190f99c5e7193472c9bfe/CODEOWNERS#L34

If you mind about this, we should improve the mappings.

File: dub.sdl
Comment (wilzbach) [inline dub.sdl]: Same change required as inhttps://github.com/dlang/dmd/pull/7415 (atm moment the DUB file is outdated).

File: dub.sdl
Comment (wilzbach) [inline dub.sdl]: FWIW splitting up dmd in four packages creates a noticeable overhead in terms of DUB compile times (compared to one big library file).

File: src/ddmd/root/rmem.d
Comment (wilzbach) [inline src/ddmd/root/rmem.d]: It seems like `version(GC)` isn't used actively (at least by anything that is part of the testsuite/CIs).

File: test/dub_package/source/app.d
Comment (wilzbach) [inline test/dub_package/source/app.d]: Is there any chance we can move a generic initialization function into the DMD codebase?

File: test/dub_package/source/app.d
Comment (wilzbach) [inline test/dub_package/source/app.d]: This is the worst bit about this PR. While, for this test we could mock `object.d`, I think we should show potential users how they can add import paths as they most likely want to do so.
Ideas:
- mock `object.d`
- use sth. like `../../../druntime/import` (i.e. depend on existing repo layout)
- call DMD and search config strings
- refactor config parsing from `mars.d`, s.t. it can be used here

File: test/dub_package/source/app.d
Comment (wilzbach) [inline test/dub_package/source/app.d]: We still need to define all predefined versions, but `addDefaultVersionIdentifiers` is `private`.
Imho if I would be about to write an IDE plugin, I wouldn't want to keep this list in sync if DMD already has the logic.
For example, to import Phobos on Linux, I need this list:

```
import ddmd.cond : VersionCondition;
VersionCondition.addPredefinedGlobalIdent("DigitalMars");
VersionCondition.addPredefinedGlobalIdent("Posix");
VersionCondition.addPredefinedGlobalIdent("linux");
VersionCondition.addPredefinedGlobalIdent("ELFv1");
VersionCondition.addPredefinedGlobalIdent("LittleEndian");
VersionCondition.addPredefinedGlobalIdent("CRuntime_Glibc");
VersionCondition.addPredefinedGlobalIdent("D_Version2");
VersionCondition.addPredefinedGlobalIdent("all");
VersionCondition.addPredefinedGlobalIdent("D_InlineAsm_X86_64");
VersionCondition.addPredefinedGlobalIdent("X86_64");
VersionCondition.addPredefinedGlobalIdent("D_HardFloat");
```

So any chance we can refactor `addDefaultVersionIdentifiers` and expose it?

File: test/dub_package/source/app.d
Comment (wilzbach) [inline test/dub_package/source/app.d]: Well, obviously it would be nice if the `dmd` "library` contains a couple of high-level wrapper, s.t. a potential author only has to interact with its shiny, modern D APIs.

File: dub.sdl
Comment (jacob-carlborg) [reply dub.sdl]: I thought it would be nice to be able to only pull in the lexer, i.e. for source code highlighting. But since Dub will currently download the whole DMD package anyway it doesn't matter that much.

Although, it would be nice to have some kind of enforcement of the separation between the different phases, to avoid that the lexer suddenly starts to depend on the rest of the front end.

File: test/dub_package/source/app.d
Comment (jacob-carlborg) [reply test/dub_package/source/app.d]: Currently this is handled in `mars.d`, which contains the `main` function of DMD. This file should not be included in the Dub package, but I don't see why we can't move this to a separate function, called `initFrontEnd` which can be called both from the compiler and the Dub package. Perhaps we can add a new file with a shared constructor that is only part of the Dub package that calls this `initFrontEnd` function.

Comment (wilzbach) [reply test/dub_package/source/app.d]: > Perhaps we can add a new file with a shared constructor that is only part of the Dub package that calls this initFrontEnd function.

Yeah, I think this would be the best way to move forward. This file could also contain other convenience, modern D wrapper. How about `ddmd/libdmd.d`, `ddmd/library.d`, or `ddmd/frontend.d`?

Comment (jacob-carlborg) [reply test/dub_package/source/app.d]: I don't like `ddmd/libdmd.d`, the others are fine, I think I prefer `ddmd/frontend.d`. But I don't want to limit ourselves to this file. I think we should look at what we want to add and add it to the modules where it makes most sense. I don't mind adding several new modules and or packages if that makes more sense.

**Comment by wilzbach** [comment]
OK, so I added a rudimentary `dmd.conf` detection which is a combination from `findConfFile` and a quick regex to parse the import paths.

-> this should now work on Travis

> I don't like ddmd/libdmd.d, the others are fine, I think I prefer ddmd/frontend.d. But I don't want to limit ourselves to this file. I think we should look at what we want to add and add it to the modules where it makes most sense. I don't mind adding several new modules and/or packages if that makes more sense. I don't have any problems with having only one symbol in a module if it doesn't fit anywhere else.

-> I created `ddmd/frontend.d` and moved the initialization calls to this module to prime it.

Other points (now not displayed anymore):

- [We still need to define all predefined versions, but addDefaultVersionIdentifiers is private.](https://github.com/dlang/dmd/pull/7425#discussion_r156264693)
- "obviously it would be nice if the dmd "library` contains a couple of high-level wrapper, s.t. a potential author only has to interact with its shiny, modern D APIs."

File: dub.sdl
Comment (wilzbach) [reply dub.sdl]: Well I hope that in practice one doesn't need to constantly recompile `libdmd.a`, so it shouldn't matter too much.

File: test/dub_package/.gitignore
Comment (wilzbach) [inline test/dub_package/.gitignore]: As hopefully more examples will be added, ignoring all binary files in this directory avoids frequent updates of this file.

File: dub.sdl
Comment (jacob-carlborg) [inline dub.sdl]: The indentation is wrong here.

File: dub.sdl
Comment (jacob-carlborg) [inline dub.sdl]: The indentation is wrong here.

File: test/dub_package/README.md
Comment (jacob-carlborg) [inline test/dub_package/README.md]: This line and line [10](https://github.com/dlang/dmd/pull/7425/files#diff-16b8627c835f07289426c754bd2472efR10) doesn't have the same indentation. For code blocks I recommend the triple backticks.

File: test/dub_package/README.md
Comment (jacob-carlborg) [inline test/dub_package/README.md]: Personally I prefer the hash `#` for headers. Then I only need one character and don't have to add/remove characters if the title changes.

File: test/dub_package/frontend.d
Comment (jacob-carlborg) [inline test/dub_package/frontend.d]: Opening curly braces on their own lines.

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: test/dub_package/README.md
Comment (wilzbach) [reply test/dub_package/README.md]: Yeah I usually use a mix of these styles in my markdown: `===` for the title and `#` for all headers within the file as the title usually changes seldom, but has more emphasis this way.
Anyhow I wouldn't mind at all to use a `#` if that's important, but I think there are more pressing issues than which flavor of Markdown should be used?

**Comment by wilzbach** [comment]
(addressed the comments & hopefully found the issue for the failing Travis test.)

File: test/dub_package/README.md
Comment (jacob-carlborg) [reply test/dub_package/README.md]: No, it's not that important. That's why I said it's my personal preference.

File: dub.sdl
Comment (jacob-carlborg) [inline dub.sdl]: Minor detail, it would be nice if the list was sorted alphabetically, unless it's sorted in some other way that I fail to see now.

File: dub.sdl
Comment (jacob-carlborg) [inline dub.sdl]: The last three files, I know they're not technically not part of the backend but I think they feel more geared towards the backend or code generation. Do we want to include them?

File: src/ddmd/frontend.d
Comment (jacob-carlborg) [inline src/ddmd/frontend.d]: This is not used as far as I can see.

File: test/dub_package/frontend.d
Comment (jacob-carlborg) [inline test/dub_package/frontend.d]: Do we need to worry about this when the tests are also run on macOS?

**Review by jacob-carlborg — CHANGES_REQUESTED** [review]

File: dub.sdl
Comment (wilzbach) [reply dub.sdl]: Currently they are required by some bits of the frontend (see below). Though I'm pretty sure, that these parts could `version`ed out in the future.

#### Remove libelf.d

```
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(lib.o):(.data.rel.ro+0x28): undefined reference to `_D4ddmd6libelf12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(lib_532_2d5.o): In function `_D4ddmd3lib7Library7factoryFZCQBcQBaQz':
/home/seb/dlang/dmd/src/ddmd/lib.d:56: undefined reference to `LibElf_factory()'
collect2: error: ld returned 1 exit status
Error: linker exited with status 1
```

#### Remove `gluelayer.d`

```
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(dclass.o):(.data.rel.ro+0x20): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(denum.o):(.data.rel.ro+0x10): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(dmodule.o):(.data.rel.ro+0x28): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(dsymbol.o):(.data.rel.ro+0x30): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(dsymbolsem.o):(.data.rel.ro+0xd8): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(dsymbolsem_39a_657.o): In function `Semantic3Visitor::visit(FuncDeclaration*)':
/home/seb/dlang/dmd/src/ddmd/dsymbolsem.d:1043: undefined reference to `retStyle(TypeFunction*)'
/home/seb/dlang/dmd/src/ddmd/dsymbolsem.d:1097: undefined reference to `retStyle(TypeFunction*)'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(expression.o):(.data.rel.ro+0x40): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(func.o):(.data.rel.ro+0x28): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(mars.o):(.data.rel.ro+0x20): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(mtype.o):(.data.rel.ro+0x38): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(objc.o):(.data.rel.ro+0x60): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(objc_5ba_3c6.o): In function `_D4ddmd4objc9Supported6__ctorMFZCQBfQBdQBb':
/home/seb/dlang/dmd/src/ddmd/objc.d:185: undefined reference to `objc_initSymbols()'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(statement.o):(.data.rel.ro+0x30): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(statementsem.o):(.data.rel.ro+0xd0): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(statementsem_632_9d9.o): In function `StatementSemanticVisitor::visit(AsmStatement*)':
/home/seb/dlang/dmd/src/ddmd/statementsem.d:3923: undefined reference to `asmSemantic(AsmStatement*, Scope*)'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(typinf.o):(.data.rel.ro+0x48): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(aggregate.o):(.data.rel.ro+0x18): undefined reference to `_D4ddmd9gluelayer12__ModuleInfoZ'
/tmp/.dub/build/dmd-~dub-frontend/library-debug-linux.posix-x86_64-dmd_2077-A550EF4C39A32552F8FE470A957BC323/libdmd_frontend.a(typinf_680_468.o): In function `genTypeInfo(Type*, Scope*)':
/home/seb/dlang/dmd/src/ddmd/typinf.d:67: undefined reference to `toObjFile(Dsymbol*, bool)'
```

Comment (jacob-carlborg) [reply dub.sdl]: I see. I think we can do some refactoring later to remove these dependencies if we want to.

BTW, I think `lib.d` is for generating libraries, which I don't think we need.

**Review by jacob-carlborg — APPROVED** [review]

**Comment by jacob-carlborg** [comment]
LGTM 👌.

**Review by andralex — APPROVED** [review]
For an initial release, looks great. cc @RazvanN7

File: dub.sdl
Comment (jacob-carlborg) [reply dub.sdl]: One way is to move every file that is part of the Dub package to its own directory, then the whole directory can be included with `sourcePaths`.

Comment (wilzbach) [reply dub.sdl]: DUB also supports `excludedSourceFiles` for removing files from a `sourcePaths` set

Comment (wilzbach) [reply dub.sdl]: Give it a try. Seems to be simple enough, though it would have been easier with a better layout of DMD (see the exclude file globs) and I don't know how fast the globbing is implemented in DUB (though it's a handful of files)...

If you want to cleanup the house, some ideas:
- move everything related to the IR -> `dmd.ir`
- move everything related to reading & writing object files `src/dmd/{scan,lib}{elfmach,mscoff,omf}.d` -> `dmd.obj`

File: test/dub_package/.gitignore
Comment (ZombineDev) [reply test/dub_package/.gitignore]: I think a cleaner solution would be put all dub artifacts in `generated/` as well, using `copyFiles` (see http://code.dlang.org/package-format?lang=sdl).

**Comment by wilzbach** [comment]
> For an initial release,

So I finally got Travis to pass :)
I had to add a small configuration file searcher for LDC to get Travis to pass.
I excluded GDC from checking the dub package for now, because there were too many errors and using Phobos fully is very convenient. Also IIRC a new release for GDC is scheduled to happen sometime in early 2018, so by the time this DUB package is fully usable (i.e. we have proper semantic versioning and have worked out all the other issues), using DMD as a library via DUB with GDC won't be a problem either. From what I can tell are these examples intended to be showcases anyhow and this PR changes the status quo from not usable to compiling & working with a simple example.

Finally the "simple" frontend example now has more than 150 lines, so in a follow-up PR we should try to factor out common logic to `dmd.frontend` (CC @RazvanN7).

**Comment by wilzbach** [comment]
@andralex @jacob-carlborg @RazvanN7 (or other) okay to move forward with this as an initial version? I would like to look into @joakim-noah's [report](http://forum.dlang.org/post/vvhtgzyyofehmrwsfrlq@forum.dlang.org) that we can't due a full semantic on Phobos yet and also start to build stuff on top of this ;-)

**Comment by jacob-carlborg** [comment]
Yes, please move forward 👌.

**Comment by jacob-carlborg** [comment]
Auto-merge toggled on

File: test/dub_package/frontend.d
Comment (RazvanN7) [inline test/dub_package/frontend.d]: Why isn't this in frontend.d? (src/dmd/frontend.d)

File: test/dub_package/frontend.d
Comment (RazvanN7) [inline test/dub_package/frontend.d]: Why isn't this in frontend.d ?

File: test/dub_package/frontend.d
Comment (RazvanN7) [inline test/dub_package/frontend.d]: This should certainly be in frontend.d and we could call it  `fullSemanticAnalysis` or something along these terms.

**Review by RazvanN7 — COMMENTED** [review]
Other than some minor comments it looks good. The next step should be to expose more semantic methods in a way that is easy to use, for example: the one which I find very important is the search method which does symbol resolution.

File: test/dub_package/frontend.d
Comment (jacob-carlborg) [reply test/dub_package/frontend.d]: Unless there's a reason to not run the full semantic analysis I don't think the `full` prefix is necessary. I hope that we can keep the three phases an implementation detail.

**Comment by wilzbach** [comment]
> The next step should be to expose more semantic methods in a way that is easy to use, for example: the one which I find very important is the search method which does symbol resolution.

Yes they should be (I think I mentioned this a couple of times), but for now I didn't want to push anything to `frontend.d` as I didn't want to stall this PR unnecessarily with a discussion of the exposed user API (@RazvanN7 how about having a chat in the next days regarding this?)
Also, once this is in, we can trivially refactor the code, as we will have at least one CI checking it :)

File: test/dub_package/frontend.d
Comment (RazvanN7) [reply test/dub_package/frontend.d]: I'm not sure this is possible. As a tool developer you may want to visit the AST while it is in a certain state. Given the fact that the semantic methods modify the AST quite a lot you might want to take a look at it before a certain semantic pass is done. For example: enums are evaluated at compile time  (during semantic3) ; in order to do those modifications the semantic3 stage alters the AST; as a tool developer I may want to do something in with the original AST which is not longer possible given that semantic3 has modified it.

Comment (wilzbach) [reply test/dub_package/frontend.d]: Hehe that's exactly why I didn't want to push this into `frontend.d` for now ;-)

**Comment by RazvanN7** [comment]
@wilzbach Sure thing. I was just suggesting the next step after this PR is merged.

File: test/dub_package/frontend.d
Comment (RazvanN7) [reply test/dub_package/frontend.d]: That's not a problem. You have the individual semantic passes, but if you don't care about that you can just call `FullSemanticAnalysys` or some other name that you find suited.

Comment (jacob-carlborg) [reply test/dub_package/frontend.d]: Please let leave it as is and continue improve it after this is merged.

**Comment by wilzbach** [comment]
> Why isn't this in frontend.d? (src/dmd/frontend.d)
>  I was just suggesting the next step after this PR is merged.

Yes, let's continue the discussion at https://github.com/dlang/dmd/pull/7485

**Pull request merged**
